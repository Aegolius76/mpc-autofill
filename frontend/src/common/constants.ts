export const ProjectName = "MPC Autofill";

import { CardType, Faces } from "@/common/types";

export const Card: CardType = "CARD";
export const Cardback: CardType = "CARDBACK";
export const Token: CardType = "TOKEN";

export const CardTypeSeparator = ":";

export const CardTypePrefixes: { [prefix: string]: CardType } = {
  "": Card,
  b: Cardback,
  t: Token,
};

export const ReversedCardTypePrefixes = Object.fromEntries(
  Object.keys(CardTypePrefixes).map((prefix: string) => [
    CardTypePrefixes[prefix],
    prefix.length > 0 ? prefix + CardTypeSeparator : prefix,
  ])
);

export const Front: Faces = "front";
export const Back: Faces = "back";
export const FaceSeparator = "|";

export const ToggleButtonHeight = 38; // pixels

export const MinimumDPI = 0;
export const MaximumDPI = 1500;
export const DPIStep = 50;

export const MaximumSize = 30; // megabytes
export const SizeStep = 1;

export const CSRFKey = "csrftoken";
export const SearchSettingsKey = "searchSettings";
export const GoogleAnalyticsConsentKey = "googleAnalyticsConsent";
export const BackendURLKey = "backendURL";

export const Brackets: Array<number> = [
  18, 36, 55, 72, 90, 108, 126, 144, 162, 180, 198, 216, 234, 396, 504, 612,
];

export const ProjectMaxSize: number = Brackets[Brackets.length - 1];

export enum QueryTags {
  BackendSpecific = "backendSpecific",
}
