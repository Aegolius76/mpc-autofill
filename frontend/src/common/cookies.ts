/**
 * Data API for interacting with anything stored in cookies.
 */

import Cookies from "js-cookie";
import Ajv from "ajv";
import {
  SearchSettings,
  SourceDocument,
  SourceDocuments,
  SourceRow,
} from "./types";
import { searchSettingsSchema } from "./schemas";
import {
  CSRFCookie,
  SearchSettingsCookie,
  GoogleAnalyticsConsentCookie,
  MaximumDPI,
  MaximumSize,
  MinimumDPI,
  BackendURLCookie,
} from "./constants";

const ajv = new Ajv();

//# region CSRF
// TODO: unsure if we still need this.

export function getCSRFHeader(): HeadersInit | undefined {
  const csrfToken = Cookies.get(CSRFCookie);
  if (csrfToken != null) {
    return { "X-CSRFToken": csrfToken };
  }
  return undefined;
}

//# endregion

//# region search settings

export function getCookieSearchSettings(
  sourceDocuments: SourceDocuments
): SearchSettings {
  /**
   * Get search settings from cookie data. If valid data is retrieved,
   * ensure that all `sourceDocuments` are included in the returned settings,
   * with any new sources that weren't previously included added to the end and enabled.
   */

  const rawSettings = JSON.parse(Cookies.get(SearchSettingsCookie) ?? "{}");
  const validate = ajv.compile(searchSettingsSchema);
  const rawSettingsValid = validate(rawSettings);
  if (rawSettingsValid) {
    // reconcile against sourceDocuments
    const sourceInDatabaseSet: Set<number> = new Set(
      Object.values(sourceDocuments).map((sourceDocument) => sourceDocument.pk)
    );
    const sources: Array<SourceRow> = rawSettings.sourceSettings.sources ?? [];
    const sourceInCookieSet: Set<number> = new Set(
      sources.map((row) => row[0])
    );
    // one fat line of reconciliation, good luck reading this future nick! i wrote this at 12:26am.
    rawSettings.sourceSettings.sources = sources
      .filter((row: SourceRow) => sourceInDatabaseSet.has(row[0]))
      .concat(
        Array.from(sourceInDatabaseSet)
          .filter((pk: number) => !sourceInCookieSet.has(pk))
          .map((pk: number) => [pk, true])
      );
    return rawSettings;
  } else {
    // default settings
    return {
      searchTypeSettings: { fuzzySearch: false },
      sourceSettings: {
        sources: Object.values(sourceDocuments).map(
          (sourceDocument: SourceDocument) => [sourceDocument.pk, true]
        ),
      },
      filterSettings: {
        minimumDPI: MinimumDPI,
        maximumDPI: MaximumDPI,
        maximumSize: MaximumSize,
      },
    };
  }
}

export function setCookieSearchSettings(settings: SearchSettings): void {
  Cookies.set(SearchSettingsCookie, JSON.stringify(settings), {
    expires: 365,
    sameSite: "strict",
  });
}

//# endregion

//# region google analytics consent

export function getGoogleAnalyticsConsent(): boolean | undefined {
  const rawConsent = Cookies.get(GoogleAnalyticsConsentCookie);
  return rawConsent != undefined ? JSON.parse(rawConsent) === true : undefined;
}

export function setGoogleAnalyticsConsent(consent: boolean): void {
  Cookies.set(GoogleAnalyticsConsentCookie, JSON.stringify(consent), {
    expires: 365,
    sameSite: "strict",
  });
}

//# endregion

//# region backend

export function getCookieBackendURL(): string | undefined {
  return Cookies.get(BackendURLCookie);
}

export function setCookieBackendURL(url: string): void {
  Cookies.set(BackendURLCookie, url, {
    expires: 365,
    sameSite: "strict",
  });
}

export function clearCookieBackendURL(): void {
  Cookies.remove(BackendURLCookie);
}

//# endregion
