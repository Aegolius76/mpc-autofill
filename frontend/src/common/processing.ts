/**
 * This module contains functions which sanitise and transform user inputs for cards to query into useful formats.
 */

import {
  Card,
  CardTypePrefixes,
  CardTypeSeparator,
  FaceSeparator,
  SelectedImageSeparator,
} from "@/common/constants";
import {
  DFCPairs,
  ProcessedLine,
  ProjectMember,
  SearchQuery,
} from "@/common/types";

export function sanitiseWhitespace(text: string): string {
  /**
   * Clean any instances of doubled-up whitespace from `text`.
   */

  const re = / +(?= )/g;
  return text.replaceAll(re, "").trim();
}

export function stripTextInParentheses(text: string): string {
  /**
   * Remove all text within (parentheses) from `text`.
   * Does not handle (nested (parentheses)). TODO: update this function to do this
   */

  const re = /[([].*?[)\]]/g;
  return sanitiseWhitespace(text.replaceAll(re, ""));
}

export function processQuery(query: string): string {
  /**
   * Process `query` by converting to lowercase, removing all punctuation, and sanitising whitespace.
   * Note that hyphens are not removed due to how Elasticsearch's classic tokenizer works:
   * https://www.elastic.co/guide/en/elasticsearch/reference/7.17/analysis-classic-tokenizer.html
   */

  // TODO: remove any numbers from the front
  // escaping \[ is technically unnecessary, but I think it's more readable to escape it
  return sanitiseWhitespace(
    query
      .toLowerCase()
      .trim()
      // eslint-disable-next-line
      .replace(/[~`!@#$%^&*(){}\[\];:"'<,.>?/\\|_+=]/g, "")
  );
}

export function processPrefix(query: string): SearchQuery {
  /**
   * Identify the prefix of a query. For example, `query`="t:goblin" would yield
   *   {query: "goblin", card_type: TOKEN}.
   */

  for (const [prefix, cardType] of Object.entries(CardTypePrefixes)) {
    if (
      prefix !== "" &&
      query
        .trimStart()
        .toLowerCase()
        .startsWith(`${prefix}${CardTypeSeparator}`)
    ) {
      return {
        query: processQuery(query.trimStart().slice(prefix.length + 1)),
        card_type: cardType,
      };
    }
  }
  return { query: processQuery(query), card_type: CardTypePrefixes[""] };
}

function unpackLine(
  line: string
): [number, [string, string | null] | null, [string, string | null] | null] {
  /**
   * Unpack `line` into its constituents.
   *
   * Inputs to this function are unpacked according to the below schema. For example, consider `4x opt@1234 | char@xyz`:
   *      4x                opt        @        1234         |      char       @        xyz
   * â””â”€ quantity â”€â”€â”˜ â””â”€ front query â”€â”€â”˜ â””â”€ front image ID â”€â”€â”˜ â””â”€ back query â”€â”€â”˜ â””â”€ back image ID â”€â”€â”˜
   *
   * If quantity is not specified, we assume a quantity of 1.
   * Specifying a back query is optional.
   * Specifying an image ID (for each face) is optional.
   *
   * (sorry for jamming this much stuff into one regex ðŸ—¿)
   */

  const trimmedLine = line.replace(/\s+/g, " ").trim();
  const re = new RegExp(
    `^(?:([0-9]*)?[xX]?\\s?(.*?)(?:${SelectedImageSeparator}([A-z0-9_\\-]*))?)?(?:(?:\\s*)${
      "\\" + FaceSeparator
    }(?:\\s*)(.+?)(?:${SelectedImageSeparator}([A-z0-9_\\-]*))?)?$`,
    "gm"
  );
  const results = re.exec(trimmedLine);
  if (results == null) {
    return [0, null, null];
  }
  return [
    parseInt(results[1] ?? "1"),
    [results[2], results[3]],
    [results[4], results[5]],
  ];
}

export function processLine(line: string, dfcPairs: DFCPairs): ProcessedLine {
  /**
   * Process `line` to identify the search query and the number of instances requested for each face.
   * If no back query is specified, attempt to match the front query to a DFC pair.
   * For example, `line`="3x t:goblin" would yield:
   *   [3, {query: {query: "goblin", card_type: TOKEN, selectedImage: null}}, null].
   * Another example is `line`="3x forest | b:elf" would yield:
   *   [
   *     3,
   *     {query: {query: "forest", card_type: CARD}, selectedImage: null},
   *     {query: {query: "elf", card_type: TOKEN}, selectedImage: null},
   *   ].
   */

  const [quantity, frontRawQuery, backRawQuery] = unpackLine(line);

  let frontQuery: SearchQuery | null = null;
  let frontSelectedImage: string | undefined = undefined;
  if (frontRawQuery != null && (frontRawQuery[0] ?? "").length > 0) {
    frontQuery = processPrefix(frontRawQuery[0]);
    frontSelectedImage = frontRawQuery[1] ?? undefined;
  }

  let backQuery: SearchQuery | null = null;
  let backSelectedImage: string | undefined = undefined;
  if (backRawQuery != null && (backRawQuery[0] ?? "").length > 0) {
    backQuery = processPrefix(backRawQuery[0]);
    backSelectedImage = backRawQuery[1] ?? undefined;
  } else if (
    frontQuery != null &&
    frontQuery?.query != null &&
    frontQuery.query in dfcPairs
  ) {
    // attempt to match to DFC pair
    // TODO: is it problematic to assume that all DFC pairs are the `Card` type?
    backQuery = { query: dfcPairs[frontQuery.query], card_type: Card };
  }

  return [
    quantity,
    frontQuery != null
      ? { query: frontQuery, selectedImage: frontSelectedImage }
      : null,
    backQuery != null
      ? { query: backQuery, selectedImage: backSelectedImage }
      : null,
  ];
}

export function processLines(
  lines: string,
  dfcPairs: DFCPairs
): Array<ProcessedLine> {
  /**
   * Process each line in `lines`, ignoring any lines which don't contain relevant information.
   */

  const queries: Array<[number, ProjectMember | null, ProjectMember | null]> =
    [];
  lines.split(/\r?\n|\r|\n/g).forEach((line: string) => {
    if (line != null && line.trim().length > 0) {
      const [quantity, frontMember, backMember] = processLine(line, dfcPairs);
      if (quantity > 0 && (frontMember != null || backMember != null)) {
        queries.push([quantity, frontMember, backMember]);
      }
    }
  });
  return queries;
}

export function standardiseURL(url: string): string {
  /**
   * Standardise `url` in the following ways:
   * 1. Ensure a http prefix is included, defaulting to `https://` if not specified
   * 2. Trim any trailing slash and path
   */

  const re = [...url.matchAll(/^(https?:\/\/)?(.*?)(?:\/.*)?$/gm)][0];
  return (re[1] ?? "https://") + re[2];
}

// TODO: delete this when remaining API interactions have been moved to RTK query
export function formatURL(backendURL: string, routeURL: string): string {
  // TODO: implement this properly
  return backendURL + routeURL;
}
