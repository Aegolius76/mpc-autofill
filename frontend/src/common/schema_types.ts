/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export interface SearchQuery {
  query: string | null;
  card_type: "CARD" | "CARDBACK" | "TOKEN";
}
/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

/**
 * @minItems 2
 * @maxItems 2
 */
export type SourceRow = [number, boolean];

export interface SearchSettings {
  searchTypeSettings: SearchTypeSettings;
  sourceSettings: SourceSettings;
  filterSettings: FilterSettings;
}
export interface SearchTypeSettings {
  /**
   * Whether fuzzy search is enabled
   */
  fuzzySearch: boolean;
  /**
   * Whether search settings apply to cardbacks or not
   */
  filterCardbacks: boolean;
}
export interface SourceSettings {
  /**
   * The list of sources in the order they should be searched
   */
  sources: SourceRow[] | null;
}
export interface FilterSettings {
  /**
   * The minimum DPI that cards must meet to be included in search results
   */
  minimumDPI: number;
  /**
   * The maximum DPI that cards can have to be included in search results
   */
  maximumDPI: number;
  /**
   * The maximum filesize that cards can have to be included in search results
   */
  maximumSize: number;
}
