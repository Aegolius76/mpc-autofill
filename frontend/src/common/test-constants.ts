/**
 * Some pre-built objects which can be used to build up Redux state for tests.
 */

import { Card, MaximumDPI, MaximumSize, MinimumDPI } from "@/common/constants";
import {
  BackendState,
  CardDocument,
  CardDocumentsState,
  Project,
  SearchResultsState,
  SearchSettings,
  SourceDocument,
  SourceDocuments,
} from "@/common/types";

//# region backend

export const localBackendURL = "https://127.0.0.1:8000";
export const localBackend: BackendState = { url: localBackendURL };
export const noBackend: BackendState = { url: null };

//# endregion

//# region sources

export const sourceDocument1: SourceDocument = {
  pk: 0,
  key: "source_1",
  name: "Source 1",
  identifier: "id_1",
  source_type: "Google Drive",
  external_link: null,
  description: "",
};

export const sourceDocument2: SourceDocument = {
  pk: 1,
  key: "source_2",
  name: "Source 2",
  identifier: "id_2",
  source_type: "Google Drive",
  external_link: null,
  description: "",
};

export const sourceDocument3: SourceDocument = {
  pk: 2,
  key: "source_3",
  name: "Source 3",
  identifier: "id_3",
  source_type: "Google Drive",
  external_link: null,
  description: "",
};

export const sourceDocument4: SourceDocument = {
  pk: 3,
  key: "source_4",
  name: "Source 4",
  identifier: "id_4",
  source_type: "Google Drive",
  external_link: null,
  description: "",
};

export const sourceDocuments: SourceDocuments = {
  [sourceDocument1.pk]: sourceDocument1,
  [sourceDocument2.pk]: sourceDocument2,
  [sourceDocument3.pk]: sourceDocument3,
  [sourceDocument4.pk]: sourceDocument4,
};

//# endregion

//# region cards

export const cardDocument1: CardDocument = {
  identifier: "abc123",
  card_type: "CARD",
  name: "Card 1",
  priority: 0,
  source: sourceDocument1.key,
  source_name: sourceDocument1.name,
  source_id: sourceDocument1.pk,
  source_verbose: sourceDocument1.name,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 1",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument2: CardDocument = {
  identifier: "abc1234",
  card_type: "CARD",
  name: "Card 2",
  priority: 0,
  source: sourceDocument1.key,
  source_name: sourceDocument1.name,
  source_id: sourceDocument1.pk,
  source_verbose: sourceDocument1.name,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 2",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument3: CardDocument = {
  identifier: "abc12345",
  card_type: "CARD",
  name: "Card 3",
  priority: 0,
  source: sourceDocument1.key,
  source_name: sourceDocument1.name,
  source_id: sourceDocument1.pk,
  source_verbose: sourceDocument1.name,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 3",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument4: CardDocument = {
  identifier: "abc123456",
  card_type: "CARD",
  name: "Card 4",
  priority: 0,
  source: sourceDocument1.key,
  source_name: sourceDocument1.name,
  source_id: sourceDocument1.pk,
  source_verbose: sourceDocument1.name,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 4",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument5: CardDocument = {
  identifier: "abc1234567",
  card_type: "CARDBACK",
  name: "Card 5",
  priority: 0,
  source: sourceDocument2.key,
  source_name: sourceDocument2.name,
  source_id: sourceDocument2.pk,
  source_verbose: `${sourceDocument2.name} Cardbacks`,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 5",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument6: CardDocument = {
  identifier: "abc12345678",
  card_type: "TOKEN",
  name: "Card 6",
  priority: 0,
  source: sourceDocument3.key,
  source_name: sourceDocument3.name,
  source_id: sourceDocument3.pk,
  source_verbose: `${sourceDocument3.name} Tokens`,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 6",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

//# endregion

//# region project

export const projectSelectedImage1: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

export const projectThreeMembersSelectedImage1: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

export const projectSelectedImage2: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument2.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

//# endregion

//# region search settings

export const defaultSettings: SearchSettings = {
  searchTypeSettings: { fuzzySearch: false, filterCardbacks: false },
  sourceSettings: {
    sources: [
      [sourceDocument1.pk, true],
      [sourceDocument2.pk, true],
      [sourceDocument3.pk, true],
      [sourceDocument4.pk, true],
    ],
  },
  filterSettings: {
    minimumDPI: MinimumDPI,
    maximumDPI: MaximumDPI,
    maximumSize: MaximumSize,
    languages: [],
    includesTags: [],
    excludesTags: ["NSFW"],
  },
};

//# endregion
