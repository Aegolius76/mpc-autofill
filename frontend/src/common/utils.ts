import { Brackets } from "./constants";

export function wrapIndex(index: number, count: number): number {
  return ((index % count) + count) % count;
}

export function imageSizeToMBString(
  size: number,
  numDecimalPlaces: number
): string {
  /**
   * Format `size` (a size in bytes) as a string in megabytes.
   */

  const roundFactor = 10 ** numDecimalPlaces;
  let sizeMB = size / 1000000;
  sizeMB = Math.round(sizeMB * roundFactor) / roundFactor;
  return `${sizeMB} MB`;
}

export function bracket(projectSize: number): number {
  /**
   * Calculate the MPC bracket that `projectSize` falls into.
   */

  for (const bracket of Brackets) {
    if (bracket >= projectSize) {
      return bracket;
    }
  }
  return Brackets[Brackets.length - 1];
}

export function downloadImage(imageURL: string, new_tab = true) {
  /**
   * Download an image with the given download link.
   * Note that this only works when `imageURL` has the same origin as where the link was opened from.
   */

  const element = document.createElement("a");
  element.href = imageURL;
  element.setAttribute("download", "deez.png"); // TODO: can we affect file name?
  if (new_tab) element.target = "_blank";

  element.style.display = "none";
  document.body.appendChild(element);
  element.click();
  document.body.removeChild(element);
}

export function downloadText(filename: string, text: string) {
  const element = document.createElement("a");
  element.setAttribute(
    "href",
    "data:text/plain;charset=utf-8," + encodeURIComponent(text)
  );
  element.setAttribute("download", filename);

  element.style.display = "none";
  document.body.appendChild(element);

  element.click();

  document.body.removeChild(element);
}
