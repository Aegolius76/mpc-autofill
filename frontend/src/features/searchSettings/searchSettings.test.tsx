//# region snapshot tests

import { waitFor, within } from "@testing-library/dom";

import App from "@/app/app";
import { Front } from "@/common/constants";
import {
  cardDocument1,
  localBackend,
  sourceDocument1,
} from "@/common/test-constants";
import {
  expectCardGridSlotState,
  importText,
  openSearchSettingsModal,
  renderWithProviders,
} from "@/common/test-utils";
import {
  cardbacksOneResult,
  cardDocumentsOneResult,
  cardDocumentsSixResults,
  defaultHandlers,
  searchResultsOneResult,
  searchResultsSixResults,
  sourceDocumentsThreeResults,
} from "@/mocks/handlers";
import { server } from "@/mocks/server";

test("the html structure of search settings", async () => {
  server.use(
    cardDocumentsOneResult,
    cardbacksOneResult,
    sourceDocumentsThreeResults,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: { members: [], cardback: null },
    },
  });

  // we need to wait for a second here after rendering to ensure sources have been fetched
  // searching for a card here isn't strictly necessary for the snapshot test,
  // but it's a convenient way to ensure that sources have been fetched
  await importText("my search query");
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);

  const searchSettings = await openSearchSettingsModal();
  await waitFor(() => within(searchSettings).getByText(sourceDocument1.name));
  expect(searchSettings).toMatchSnapshot();
});

//# endregion
