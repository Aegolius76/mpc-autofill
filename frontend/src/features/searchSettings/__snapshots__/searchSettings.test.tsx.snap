// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of search settings 1`] = `
<div
  class="modal-dialog"
  data-testid="search-settings"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Search Settings
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <h5>
        Search Type
      </h5>
      Configure how closely the search results should match your query.
      <br />
      <br />
      <div
        class="btn toggle btn-md off btn-info"
        role="button"
        style="width: 100%; height: 38px;"
      >
        <div
          class="toggle-group"
        >
          <span
            class="btn toggle-on btn-md flex-centre btn-success"
          >
            Fuzzy (Forgiving) Search
          </span>
          <span
            class="btn toggle-off btn-md flex-centre btn-info"
          >
            Precise Search
          </span>
          <span
            class="toggle-handle btn btn-md btn-default"
          />
        </div>
      </div>
      <hr />
      <h5>
        Filters
      </h5>
      Configure the DPI (dots per inch) and file size ranges the search results must be within.
      <br />
      At a fixed physical size, a higher DPI yields a higher resolution print.
      <br />
      MakePlayingCards prints cards up to 
      <b>
        800 DPI
      </b>
      , meaning an 800 DPI print and a 1200 DPI print will 
      <b>
        look the same
      </b>
      .
      <br />
      <br />
      <div
        class="row"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
          >
            Minimum: 
            <b>
              0
               DPI
            </b>
          </label>
          <input
            class="form-range"
            max="1500"
            min="0"
            step="50"
            type="range"
            value="0"
          />
        </div>
        <div
          class="col-6"
        >
          <label
            class="form-label"
          >
            Maximum: 
            <b>
              1500
               DPI
            </b>
          </label>
          <input
            class="form-range"
            max="1500"
            min="0"
            step="50"
            type="range"
            value="1500"
          />
        </div>
      </div>
      <label
        class="form-label"
      >
        File size: Up to 
        <b>
          30
           MB
        </b>
      </label>
      <input
        class="form-range"
        max="30"
        min="0"
        step="1"
        type="range"
        value="30"
      />
      <hr />
      <h5>
        Sources
      </h5>
      Configure the sources to include in the search results.
      <ul>
        <li>
          <b>
            Drag & drop
          </b>
           them to change the order they're searched in.
        </li>
        <li>
          Use the 
          <b>
            arrows
          </b>
           to send a source to the top or bottom.
        </li>
        <li>
          Click the 
          <b>
            table header
          </b>
           to enable or disable all sources.
        </li>
      </ul>
      <div
        style="height: 215px;"
      >
        <table
          class="table"
          style="table-layout: auto;"
        >
          <thead>
            <tr
              style="height: 38px;"
            >
              <th
                class="prevent-select"
              >
                Enabled
              </th>
              <th
                class="prevent-select"
              >
                Source Name
              </th>
              <th
                class="prevent-select"
              >
                Source Type
              </th>
              <th />
              <th />
            </tr>
          </thead>
          <tbody>
            <tr
              aria-describedby="rfd-hidden-text-:r0:-hidden-text-:r1:"
              data-rfd-drag-handle-context-id=":r0:"
              data-rfd-drag-handle-draggable-id="0"
              data-rfd-draggable-context-id=":r0:"
              data-rfd-draggable-id="0"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <td
                style="vertical-align: middle; width: 20%;"
              >
                <div
                  class="btn toggle btn-md btn-primary"
                  role="button"
                  style="width: 100%; height: 38px;"
                >
                  <div
                    class="toggle-group"
                  >
                    <span
                      class="btn toggle-on btn-md flex-centre prevent-select btn-primary"
                    >
                      On
                    </span>
                    <span
                      class="btn toggle-off btn-md flex-centre prevent-select btn-secondary"
                    >
                      Off
                    </span>
                    <span
                      class="toggle-handle btn btn-md btn-default"
                    />
                  </div>
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 40%;"
              >
                <a>
                  Source 1
                </a>
              </td>
              <td
                class="prevent-select"
                style="vertical-align: middle; width: 30%;"
              >
                gdrive
              </td>
              <td
                style="vertical-align: middle; width: 5%; text-align: center;"
              >
                <div>
                  <i
                    class="bi bi-chevron-double-up"
                    style="font-size: 1em; cursor: pointer;"
                  />
                </div>
                <div>
                  <i
                    class="bi bi-chevron-double-down"
                    style="font-size: 1em; cursor: pointer;"
                  />
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 5%; text-align: center;"
              >
                <i
                  class="bi bi-grip-horizontal"
                  style="font-size: 2em;"
                />
              </td>
            </tr>
            <tr
              aria-describedby="rfd-hidden-text-:r0:-hidden-text-:r1:"
              data-rfd-drag-handle-context-id=":r0:"
              data-rfd-drag-handle-draggable-id="1"
              data-rfd-draggable-context-id=":r0:"
              data-rfd-draggable-id="1"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <td
                style="vertical-align: middle; width: 20%;"
              >
                <div
                  class="btn toggle btn-md btn-primary"
                  role="button"
                  style="width: 100%; height: 38px;"
                >
                  <div
                    class="toggle-group"
                  >
                    <span
                      class="btn toggle-on btn-md flex-centre prevent-select btn-primary"
                    >
                      On
                    </span>
                    <span
                      class="btn toggle-off btn-md flex-centre prevent-select btn-secondary"
                    >
                      Off
                    </span>
                    <span
                      class="toggle-handle btn btn-md btn-default"
                    />
                  </div>
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 40%;"
              >
                <a>
                  Source 2
                </a>
              </td>
              <td
                class="prevent-select"
                style="vertical-align: middle; width: 30%;"
              >
                gdrive
              </td>
              <td
                style="vertical-align: middle; width: 5%; text-align: center;"
              >
                <div>
                  <i
                    class="bi bi-chevron-double-up"
                    style="font-size: 1em; cursor: pointer;"
                  />
                </div>
                <div>
                  <i
                    class="bi bi-chevron-double-down"
                    style="font-size: 1em; cursor: pointer;"
                  />
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 5%; text-align: center;"
              >
                <i
                  class="bi bi-grip-horizontal"
                  style="font-size: 2em;"
                />
              </td>
            </tr>
            <tr
              aria-describedby="rfd-hidden-text-:r0:-hidden-text-:r1:"
              data-rfd-drag-handle-context-id=":r0:"
              data-rfd-drag-handle-draggable-id="2"
              data-rfd-draggable-context-id=":r0:"
              data-rfd-draggable-id="2"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <td
                style="vertical-align: middle; width: 20%;"
              >
                <div
                  class="btn toggle btn-md btn-primary"
                  role="button"
                  style="width: 100%; height: 38px;"
                >
                  <div
                    class="toggle-group"
                  >
                    <span
                      class="btn toggle-on btn-md flex-centre prevent-select btn-primary"
                    >
                      On
                    </span>
                    <span
                      class="btn toggle-off btn-md flex-centre prevent-select btn-secondary"
                    >
                      Off
                    </span>
                    <span
                      class="toggle-handle btn btn-md btn-default"
                    />
                  </div>
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 40%;"
              >
                <a>
                  Source 3
                </a>
              </td>
              <td
                class="prevent-select"
                style="vertical-align: middle; width: 30%;"
              >
                gdrive
              </td>
              <td
                style="vertical-align: middle; width: 5%; text-align: center;"
              >
                <div>
                  <i
                    class="bi bi-chevron-double-up"
                    style="font-size: 1em; cursor: pointer;"
                  />
                </div>
                <div>
                  <i
                    class="bi bi-chevron-double-down"
                    style="font-size: 1em; cursor: pointer;"
                  />
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 5%; text-align: center;"
              >
                <i
                  class="bi bi-grip-horizontal"
                  style="font-size: 2em;"
                />
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        Close Without Saving
      </button>
      <button
        class="btn btn-primary"
        type="button"
      >
        Save Changes
      </button>
    </div>
  </div>
</div>
`;
